File: protocols/proven-opcua/ffi/zig/src/opcua.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// opcua.zig -- Zig FFI implementation of proven-opcua.
//
// Implements an OPC UA (Industrial IoT) session state machine with:
//   - 64-slot mutex-protected session pool
//   - Address space node tracking (max 256 nodes per session)
//   - Subscription management (max 16 subscriptions per session)
//   - Security mode enforcement per session
//   - Service request validation per session state
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching OPCUAABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching OPCUAABI.Types.idr tag assignments)
// =========================================================================

/// OPC UA service types (ABI tags 0-10).
pub const ServiceType = enum(u8) {
    read = 0,
    write = 1,
    browse = 2,
    subscribe = 3,
    publish = 4,
    call = 5,
    create_session = 6,
    activate_session = 7,
    close_session = 8,
    create_subscription = 9,
    delete_subscription = 10,
};

/// Node classes (ABI tags 0-7).
pub const NodeClass = enum(u8) {
    object = 0,
    variable = 1,
    method = 2,
    object_type = 3,
    variable_type = 4,
    reference_type = 5,
    data_type = 6,
    view = 7,
};

/// Status codes (ABI tags 0-11).
pub const StatusCode = enum(u8) {
    good = 0,
    uncertain = 1,
    bad = 2,
    bad_node_id_unknown = 3,
    bad_attribute_id_invalid = 4,
    bad_not_readable = 5,
    bad_not_writable = 6,
    bad_out_of_range = 7,
    bad_type_mismatch = 8,
    bad_session_id_invalid = 9,
    bad_subscription_id_invalid = 10,
    bad_timeout = 11,
};

/// Security modes (ABI tags 0-2).
pub const SecurityMode = enum(u8) {
    none = 0,
    sign = 1,
    sign_and_encrypt = 2,
};

/// Session lifecycle states (ABI tags 0-5).
pub const SessionState = enum(u8) {
    idle = 0,
    connected = 1,
    created = 2,
    activated = 3,
    monitoring = 4,
    closing = 5,
};

// =========================================================================
// Internal data structures
// =========================================================================

const MAX_SESSIONS: usize = 64;
const MAX_NODES: usize = 256;
const MAX_SUBSCRIPTIONS: usize = 16;
const MAX_NAME_LEN: usize = 256;
const MAX_ENDPOINT_LEN: usize = 256;
const MAX_VALUE_LEN: usize = 256;

/// An address space node.
const NodeEntry = struct {
    node_id: u32,
    node_class: NodeClass,
    name: [MAX_NAME_LEN]u8,
    name_len: u32,
    /// Simple value storage for Variable nodes.
    value: [MAX_VALUE_LEN]u8,
    value_len: u32,
    active: bool,
};

/// A subscription entry.
const SubscriptionEntry = struct {
    sub_id: u32,
    interval_ms: u32,
    active: bool,
};

const empty_node: NodeEntry = .{
    .node_id = 0,
    .node_class = .object,
    .name = [_]u8{0} ** MAX_NAME_LEN,
    .name_len = 0,
    .value = [_]u8{0} ** MAX_VALUE_LEN,
    .value_len = 0,
    .active = false,
};

const empty_subscription: SubscriptionEntry = .{
    .sub_id = 0,
    .interval_ms = 1000,
    .active = false,
};

/// An OPC UA session.
const Session = struct {
    state: SessionState,
    endpoint: [MAX_ENDPOINT_LEN]u8,
    endpoint_len: u32,
    security_mode: SecurityMode,
    nodes: [MAX_NODES]NodeEntry,
    node_count: u32,
    subscriptions: [MAX_SUBSCRIPTIONS]SubscriptionEntry,
    subscription_count: u32,
    next_sub_id: u32,
    active: bool,
};

const empty_session: Session = .{
    .state = .idle,
    .endpoint = [_]u8{0} ** MAX_ENDPOINT_LEN,
    .endpoint_len = 0,
    .security_mode = .none,
    .nodes = [_]NodeEntry{empty_node} ** MAX_NODES,
    .node_count = 0,
    .subscriptions = [_]SubscriptionEntry{empty_subscription} ** MAX_SUBSCRIPTIONS,
    .subscription_count = 0,
    .next_sub_id = 1,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

fn findNode(idx: usize, node_id: u32) ?usize {
    for (&sessions[idx].nodes, 0..) |*n, i| {
        if (n.active and n.node_id == node_id) return i;
    }
    return null;
}

fn findSubscription(idx: usize, sub_id: u32) ?usize {
    for (&sessions[idx].subscriptions, 0..) |*s, i| {
        if (s.active and s.sub_id == sub_id) return i;
    }
    return null;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

pub export fn opcua_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new OPC UA session. Returns slot (>=0) or -1.
/// Transitions: Idle -> Connected.
pub export fn opcua_create(
    endpoint_ptr: [*]const u8,
    endpoint_len: u32,
    security_mode: u8,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (endpoint_len == 0 or endpoint_len > MAX_ENDPOINT_LEN) return -1;
    if (security_mode > 2) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.endpoint[0..endpoint_len], endpoint_ptr[0..endpoint_len]);
            s.endpoint_len = endpoint_len;
            s.security_mode = @enumFromInt(security_mode);
            s.state = .connected;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

pub export fn opcua_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

pub export fn opcua_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Create a session (OPC UA CreateSession service). Returns 0 on success.
/// Transitions: Connected -> Created.
pub export fn opcua_create_session(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connected) return 1;

    sessions[idx].state = .created;
    return 0;
}

/// Activate the session. Returns 0 on success.
/// Transitions: Created -> Activated.
pub export fn opcua_activate_session(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .created) return 1;

    sessions[idx].state = .activated;
    return 0;
}

/// Read a node attribute. Returns 0 on success, 1 on rejection.
pub export fn opcua_read_node(slot: c_int, node_id: u32, attr_id: u32) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    _ = attr_id;

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .activated and sessions[idx].state != .monitoring) return 1;
    if (findNode(idx, node_id) == null) return 1;

    return 0;
}

/// Write a node attribute. Returns 0 on success, 1 on rejection.
pub export fn opcua_write_node(
    slot: c_int,
    node_id: u32,
    attr_id: u32,
    value_ptr: [*]const u8,
    value_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    _ = attr_id;

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .activated and sessions[idx].state != .monitoring) return 1;
    if (value_len > MAX_VALUE_LEN) return 1;

    const ni = findNode(idx, node_id) orelse return 1;
    if (sessions[idx].nodes[ni].node_class != .variable) return 1;

    if (value_len > 0) {
        @memcpy(sessions[idx].nodes[ni].value[0..value_len], value_ptr[0..value_len]);
    }
    sessions[idx].nodes[ni].value_len = value_len;
    return 0;
}

/// Browse a node (simulated). Returns 0 on success, 1 if node not found.
pub export fn opcua_browse(slot: c_int, node_id: u32) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .activated and sessions[idx].state != .monitoring) return 1;
    if (findNode(idx, node_id) == null) return 1;

    return 0;
}

/// Add a node to the address space. Returns 0 on success, 1 on rejection.
pub export fn opcua_add_node(
    slot: c_int,
    node_id: u32,
    node_class: u8,
    name_ptr: [*]const u8,
    name_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .activated and sessions[idx].state != .monitoring) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    if (node_class > 7) return 1;

    // Check for duplicate node_id
    if (findNode(idx, node_id) != null) return 1;

    for (&sessions[idx].nodes) |*n| {
        if (!n.active) {
            n.* = empty_node;
            n.node_id = node_id;
            n.node_class = @enumFromInt(node_class);
            @memcpy(n.name[0..name_len], name_ptr[0..name_len]);
            n.name_len = name_len;
            n.active = true;
            sessions[idx].node_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Create a subscription. Returns 0 on success, 1 on rejection.
/// Transitions: Activated -> Monitoring.
pub export fn opcua_create_subscription(slot: c_int, interval_ms: u32) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .activated and sessions[idx].state != .monitoring) return 1;

    for (&sessions[idx].subscriptions) |*sub| {
        if (!sub.active) {
            sub.sub_id = sessions[idx].next_sub_id;
            sessions[idx].next_sub_id += 1;
            sub.interval_ms = if (interval_ms == 0) 1000 else interval_ms;
            sub.active = true;
            sessions[idx].subscription_count += 1;
            sessions[idx].state = .monitoring;
            return 0;
        }
    }
    return 1;
}

/// Delete a subscription. Returns 0 on success, 1 on rejection.
/// May transition: Monitoring -> Activated if last subscription.
pub export fn opcua_delete_subscription(slot: c_int, sub_id: u32) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const si = findSubscription(idx, sub_id) orelse return 1;

    sessions[idx].subscriptions[si].active = false;
    sessions[idx].subscription_count -= 1;

    if (sessions[idx].subscription_count == 0 and sessions[idx].state == .monitoring) {
        sessions[idx].state = .activated;
    }
    return 0;
}

/// Returns the subscription count.
pub export fn opcua_subscription_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].subscription_count;
}

/// Returns the node count.
pub export fn opcua_node_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].node_count;
}

/// Returns the security mode tag.
pub export fn opcua_get_security_mode(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].security_mode);
}

/// Close the session. Returns 0 on success, 1 on rejection.
pub export fn opcua_close(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .connected or state == .created or
        state == .activated or state == .monitoring)
    {
        sessions[idx].state = .closing;
        return 0;
    }
    return 1;
}

/// Complete cleanup. Transitions: Closing -> Idle.
pub export fn opcua_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .closing) return 1;

    sessions[idx].state = .idle;
    sessions[idx].nodes = [_]NodeEntry{empty_node} ** MAX_NODES;
    sessions[idx].node_count = 0;
    sessions[idx].subscriptions = [_]SubscriptionEntry{empty_subscription} ** MAX_SUBSCRIPTIONS;
    sessions[idx].subscription_count = 0;
    return 0;
}

/// Check if a session state transition is valid.
pub export fn opcua_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Connected
    if (from == 1 and to == 2) return 1; // Connected -> Created
    if (from == 2 and to == 3) return 1; // Created -> Activated
    if (from == 3 and to == 4) return 1; // Activated -> Monitoring
    if (from == 4 and to == 4) return 1; // Monitoring -> Monitoring
    if (from == 4 and to == 3) return 1; // Monitoring -> Activated
    if (from == 1 and to == 5) return 1; // Connected -> Closing
    if (from == 2 and to == 5) return 1; // Created -> Closing
    if (from == 3 and to == 5) return 1; // Activated -> Closing
    if (from == 4 and to == 5) return 1; // Monitoring -> Closing
    if (from == 5 and to == 0) return 1; // Closing -> Idle
    return 0;
}

File: bindings/elixir/lib/proven_servers/opcua.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Opcua do
  @moduledoc """
  OPC UA (OPC Unified Architecture) types for the proven-servers ABI.
  
  Mirrors the Idris2 module `OPCUAABI.Types` and its type definitions:
  - `ServiceType`   — OPC UA service types (11 constructors, tags 0-10)
  - `NodeClass`     — OPC UA node classes (8 constructors, tags 0-7)
  - `StatusCode`    — OPC UA status codes (12 constructors, tags 0-11)
  - `SecurityMode`  — Message security modes (3 constructors, tags 0-2)
  - `SessionState`  — OPC UA session lifecycle (6 constructors, tags 0-5)
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard OPC UA TCP port."
  @spec opcua_port() :: non_neg_integer()
  def opcua_port, do: 4840

  @doc "Standard OPC UA TCP/TLS port."
  @spec opcua_tls_port() :: non_neg_integer()
  def opcua_tls_port, do: 4843

  # ===========================================================================
  # ServiceType (tags 0-10)
  # ===========================================================================

  @typedoc """
  ServiceType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type service_type ::
          :read
          | :write
          | :browse
          | :subscribe
          | :publish
          | :call
          | :create_session
          | :activate_session
          | :close_session
          | :create_subscription
          | :delete_subscription

  @service_type_tags %{
    read: 0,
    write: 1,
    browse: 2,
    subscribe: 3,
    publish: 4,
    call: 5,
    create_session: 6,
    activate_session: 7,
    close_session: 8,
    create_subscription: 9,
    delete_subscription: 10,
  }

  @tag_to_service_type Map.new(@service_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ServiceType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..10, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Opcua.service_type_from_tag(0)
      {:ok, :read}
  """
  @spec service_type_from_tag(non_neg_integer()) :: {:ok, service_type()} | :error
  def service_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 10 do
    {:ok, Map.fetch!(@tag_to_service_type, tag)}
  end

  def service_type_from_tag(_tag), do: :error

  @doc """
  Encode a `ServiceType` to the C-ABI tag value.
  """
  @spec service_type_to_tag(service_type()) :: non_neg_integer()
  def service_type_to_tag(val) when is_map_key(@service_type_tags, val) do
    Map.fetch!(@service_type_tags, val)
  end

  @doc """
  All `ServiceType` variants in tag order.
  """
  @spec all_service_types() :: [service_type()]
  def all_service_types do
    [
      :read, :write, :browse, :subscribe, :publish, :call, :create_session,
      :activate_session, :close_session, :create_subscription, :delete_subscription,
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this service modifies server state.
  """
  @spec is_write?(service_type()) :: boolean()
  def is_write?(val) when val in [:write, :call], do: true
  def is_write?(_val), do: false

  @doc """
  Whether this service is a session management operation.
  """
  @spec is_session_management?(service_type()) :: boolean()
  def is_session_management?(val) when val in [:create_session, :activate_session, :close_session], do: true
  def is_session_management?(_val), do: false

  @doc """
  Whether this service relates to subscriptions.
  """
  @spec is_subscription_related?(service_type()) :: boolean()
  def is_subscription_related?(val) when val in [:subscribe, :publish, :create_subscription, :delete_subscription], do: true
  def is_subscription_related?(_val), do: false

  # ===========================================================================
  # NodeClass (tags 0-7)
  # ===========================================================================

  @typedoc """
  NodeClass types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type node_class ::
          :object
          | :variable
          | :method
          | :object_type
          | :variable_type
          | :reference_type
          | :data_type
          | :view

  @node_class_tags %{
    object: 0,
    variable: 1,
    method: 2,
    object_type: 3,
    variable_type: 4,
    reference_type: 5,
    data_type: 6,
    view: 7,
  }

  @tag_to_node_class Map.new(@node_class_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `NodeClass` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..7, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Opcua.node_class_from_tag(0)
      {:ok, :object}
  """
  @spec node_class_from_tag(non_neg_integer()) :: {:ok, node_class()} | :error
  def node_class_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 7 do
    {:ok, Map.fetch!(@tag_to_node_class, tag)}
  end

  def node_class_from_tag(_tag), do: :error

  @doc """
  Encode a `NodeClass` to the C-ABI tag value.
  """
  @spec node_class_to_tag(node_class()) :: non_neg_integer()
  def node_class_to_tag(val) when is_map_key(@node_class_tags, val) do
    Map.fetch!(@node_class_tags, val)
  end

  @doc """
  All `NodeClass` variants in tag order.
  """
  @spec all_node_classs() :: [node_class()]
  def all_node_classs do
    [
      :object, :variable, :method, :object_type, :variable_type, :reference_type,
      :data_type, :view
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this node class is an instance node (not a type definition).
  """
  @spec is_instance?(node_class()) :: boolean()
  def is_instance?(val) when val in [:object, :variable, :method, :view], do: true
  def is_instance?(_val), do: false

  @doc """
  Whether this node class is a type definition.
  """
  @spec is_type?(node_class()) :: boolean()
  def is_type?(val) when val in [:object_type, :variable_type, :reference_type, :data_type], do: true
  def is_type?(_val), do: false

  # ===========================================================================
  # StatusCode (tags 0-11)
  # ===========================================================================

  @typedoc """
  StatusCode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type status_code ::
          :good
          | :uncertain
          | :bad
          | :bad_node_id_unknown
          | :bad_attribute_id_invalid
          | :bad_not_readable
          | :bad_not_writable
          | :bad_out_of_range
          | :bad_type_mismatch
          | :bad_session_id_invalid
          | :bad_subscription_id_invalid
          | :bad_timeout

  @status_code_tags %{
    good: 0,
    uncertain: 1,
    bad: 2,
    bad_node_id_unknown: 3,
    bad_attribute_id_invalid: 4,
    bad_not_readable: 5,
    bad_not_writable: 6,
    bad_out_of_range: 7,
    bad_type_mismatch: 8,
    bad_session_id_invalid: 9,
    bad_subscription_id_invalid: 10,
    bad_timeout: 11,
  }

  @tag_to_status_code Map.new(@status_code_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `StatusCode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..11, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Opcua.status_code_from_tag(0)
      {:ok, :good}
  """
  @spec status_code_from_tag(non_neg_integer()) :: {:ok, status_code()} | :error
  def status_code_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 11 do
    {:ok, Map.fetch!(@tag_to_status_code, tag)}
  end

  def status_code_from_tag(_tag), do: :error

  @doc """
  Encode a `StatusCode` to the C-ABI tag value.
  """
  @spec status_code_to_tag(status_code()) :: non_neg_integer()
  def status_code_to_tag(val) when is_map_key(@status_code_tags, val) do
    Map.fetch!(@status_code_tags, val)
  end

  @doc """
  All `StatusCode` variants in tag order.
  """
  @spec all_status_codes() :: [status_code()]
  def all_status_codes do
    [
      :good, :uncertain, :bad, :bad_node_id_unknown, :bad_attribute_id_invalid,
      :bad_not_readable, :bad_not_writable, :bad_out_of_range, :bad_type_mismatch,
      :bad_session_id_invalid, :bad_subscription_id_invalid, :bad_timeout,
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this status code indicates success.
  """
  @spec is_good?(status_code()) :: boolean()
  def is_good?(val) when val in [:good], do: true
  def is_good?(_val), do: false

  @doc """
  Whether this status code indicates a definite failure.
  """
  @spec is_bad?(status_code()) :: boolean()
  def is_bad?(val) when val in [:good, :uncertain], do: false
  def is_bad?(_val), do: true

  @doc """
  Whether this status code relates to security/session issues.
  """
  @spec is_security_related?(status_code()) :: boolean()
  def is_security_related?(val) when val in [:bad_session_id_invalid], do: true
  def is_security_related?(_val), do: false

  # ===========================================================================
  # SecurityMode (tags 0-2)
  # ===========================================================================

  @typedoc """
  SecurityMode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type security_mode :: :none | :sign | :sign_and_encrypt

  @security_mode_tags %{
    none: 0,
    sign: 1,
    sign_and_encrypt: 2,
  }

  @tag_to_security_mode Map.new(@security_mode_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `SecurityMode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..2, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Opcua.security_mode_from_tag(0)
      {:ok, :none}
  """
  @spec security_mode_from_tag(non_neg_integer()) :: {:ok, security_mode()} | :error
  def security_mode_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 2 do
    {:ok, Map.fetch!(@tag_to_security_mode, tag)}
  end

  def security_mode_from_tag(_tag), do: :error

  @doc """
  Encode a `SecurityMode` to the C-ABI tag value.
  """
  @spec security_mode_to_tag(security_mode()) :: non_neg_integer()
  def security_mode_to_tag(val) when is_map_key(@security_mode_tags, val) do
    Map.fetch!(@security_mode_tags, val)
  end

  @doc """
  All `SecurityMode` variants in tag order.
  """
  @spec all_security_modes() :: [security_mode()]
  def all_security_modes, do: [:none, :sign, :sign_and_encrypt]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether messages are signed.
  """
  @spec is_signed?(security_mode()) :: boolean()
  def is_signed?(val) when val in [:sign, :sign_and_encrypt], do: true
  def is_signed?(_val), do: false

  @doc """
  Whether messages are encrypted.
  """
  @spec is_encrypted?(security_mode()) :: boolean()
  def is_encrypted?(val) when val in [:sign_and_encrypt], do: true
  def is_encrypted?(_val), do: false

  # ===========================================================================
  # SessionState (tags 0-5)
  # ===========================================================================

  @typedoc """
  SessionState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type session_state :: :idle | :connected | :created | :activated | :monitoring | :closing

  @session_state_tags %{
    idle: 0,
    connected: 1,
    created: 2,
    activated: 3,
    monitoring: 4,
    closing: 5,
  }

  @tag_to_session_state Map.new(@session_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `SessionState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Opcua.session_state_from_tag(0)
      {:ok, :idle}
  """
  @spec session_state_from_tag(non_neg_integer()) :: {:ok, session_state()} | :error
  def session_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_session_state, tag)}
  end

  def session_state_from_tag(_tag), do: :error

  @doc """
  Encode a `SessionState` to the C-ABI tag value.
  """
  @spec session_state_to_tag(session_state()) :: non_neg_integer()
  def session_state_to_tag(val) when is_map_key(@session_state_tags, val) do
    Map.fetch!(@session_state_tags, val)
  end

  @doc """
  All `SessionState` variants in tag order.
  """
  @spec all_session_states() :: [session_state()]
  def all_session_states, do: [:idle, :connected, :created, :activated, :monitoring, :closing]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether the session can accept service requests.
  """
  @spec can_service?(session_state()) :: boolean()
  def can_service?(val) when val in [:activated, :monitoring], do: true
  def can_service?(_val), do: false

  @doc """
  Whether the session is in a transient state.
  """
  @spec is_transient?(session_state()) :: boolean()
  def is_transient?(val) when val in [:connected, :created, :closing], do: true
  def is_transient?(_val), do: false

end
